assignments/tasks.go

```diff
@@ -0,0 +1,284 @@
+)
+
+// Things to do:
+// - Repositories in database do not have a "Name"-field.
+// - Ordering of tasks (See teacher.md)
+
+// newTask returns a task from markdown contents and associates it with the given assignment.
+// The provided markdown contents must contain a title specified on the first line,
+// starting with the "# " character sequence, followed by two new line characters.
+func newTask(contents []byte, assignment *pb.Assignment, name string) (*pb.Task, error) {
+	if !bytes.HasPrefix(contents, []byte("# ")) {
+		return nil, fmt.Errorf("task for assignment %s does not start with a # title marker", assignment.Name)
+	}
+	bodyIndex := bytes.Index(contents, []byte("\n\n"))
+	if bodyIndex == -1 {
+		return nil, fmt.Errorf("failed to find task body in %s", assignment.Name)
+	}
+
+	task := &pb.Task{
+		AssignmentOrder: assignment.GetOrder(),
+		Title:           string(contents[2:bodyIndex]),
+		Body:            string(contents[bodyIndex+2:]),
+		Name:            name,
+	}
+
+	return task, nil
+}
+
+// Updates an issue on specified repository
+func updateScmIssue(c context.Context, sc scm.SCM, course *pb.Course, repo *pb.Repository, issue *pb.Issue) (*scm.Issue, error) {
+	issueOptions := &scm.CreateIssueOptions{
+		Organization: course.Name,
+		Repository:   filepath.Base(repo.GetHTMLURL()), // Todo
+		Title:        issue.Title,
+		Body:         issue.Body,
+	}
+	return sc.EditRepoIssue(c, int(issue.IssueNumber), issueOptions)
+}
+
+// Creates an issue on specified repository.
+func createScmIssue(c context.Context, sc scm.SCM, course *pb.Course, repo *pb.Repository, task *pb.Task) (*scm.Issue, error) {
+	issueOptions := &scm.CreateIssueOptions{
+		Organization: course.Name,
+		Repository:   filepath.Base(repo.GetHTMLURL()), // Needs to be of type "tests", not "https://github.com/qf101/tests". This is a very hacky solution. pb.Repository should probably have a field "Name" that is set upon creation.
+		Title:        task.Title,
+		Body:         task.Body,
+	}
+	issue, err := sc.CreateIssue(c, issueOptions)
+	if err != nil {
+		return nil, err
+	}
+	return issue, nil
+}
+
+// This is more of a converter function, and cannot currently return an error. Should probably be renamed or something.
+func createIssue(c context.Context, repo *pb.Repository, task *pb.Task, scmIssue *scm.Issue) (*pb.Issue, error) {
+	issue := &pb.Issue{
+		RepositoryID: repo.ID,
+		IssueNumber:  uint64(scmIssue.IssueNumber),
+		Name:         task.Name,
+		Title:        task.Title,
+		Body:         task.Body,
+	}
+	return issue, nil
+}
+
+// Following is Oje code (placement might be temporary):
+
+func getTasksFromAssignments(c context.Context, assignments []*pb.Assignment) []*pb.Task {
+	tasks := []*pb.Task{}
+	for _, assignment := range assignments {
+		tasks = append(tasks, assignment.Tasks...)
+	}
+
+	return tasks
+}
+
+func handleTasks(c context.Context, db database.Database, s scm.SCM, course *pb.Course, tasks []*pb.Task) error {
```

Comment:
This is doing two different things and should be decoupled:

- Make new database method `SynchronizeTasks(course, tasks)` or `SynchronizeAssignmentTasks(course, tasks)`. This method should do the `GetAssignmentsByCourse()` and the loop over assignments and update the tasks as you do in the current `synchronizeTasks()` function below. Note that this method will not need the context.Context parameter since that's not used in the database.
- A separate method to synchronize the issues... (haven't got a specific suggestion at the moment... maybe soon, but it should only deal with updating the issues on the repos.

Reply:
By making a new `SynchronizeTasks()` database method, we can actually remove these three methods from the `Database` interface:
```go
	// CreateTasks creates slice of tasks
	CreateTasks(tasks []*pb.Task) error
	// DeleteTask deletes specified task
	DeleteTask(task *pb.Task) error
	// UpdateTasks updates slice of tasks
	UpdateTasks(tasks []*pb.Task) error
```
Each of these are used only once in connection with the sync operation. Thus, they can be made unexported and not part of the interface.